perceval/client.py

```diff
@@ -0,0 +1,196 @@
+# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources taking care of retrying requests in case connection issues. If the data source does
+    not send back a response after retrying a request, a RetryError exception is thrown.
+
+    Derivated classes can use the methods fetch and _fetch to obtain data from the data source.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param base_url: base URL of the data source
+    :param max_retries: number of max retries to a data source before raising a RetryError exception
+    :param default_sleep_time: default time to sleep in case of connection problems
+    :param rate_limit_handler: object to handle rate limit issues
+    """
+    version = '0.1'
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.session = self.create_http_session()
+        self.max_retries = max_retries
+        self.default_sleep_time = default_sleep_time
+
+    def __del__(self):
+        self.close_http_session()
+
+    def create_http_session(self, headers=None):
+        session = requests.Session()
+
+        if headers:
+            session.headers.update(headers)
+
+        return session
+
+    def close_http_session(self):
+        self.session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
+        if status_codes:
+            Retry.RETRY_AFTER_STATUS_CODES = frozenset(status_codes)
+        else:
+            Retry.RETRY_AFTER_STATUS_CODES = self.DEFAULT_RETRY_AFTER_STATUS_CODES
+
+    def fetch(self, url, payload=None, headers=None, method=GET,
+              stream=False,
+              retry_on_status_codes=None,
+              retry_after_status_code=None):
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.set_retry_after_status_codes(retry_after_status_codes)
```

Comment:
I would move all this retrying setup to `create_http_session()`.

Reply:
if we move it to _create_http_session()_, we won't be able to control every call to __fetch_. Imagine that for a given call we want to set the retry after for a 403, but in another call we don't want.

Btw, I admit I haven't found any example in the backends where this happens